### PR TITLE
updates for vLLM==0.5.4

### PR DESCRIPTION
**Pull request by dtrifiro**
adaptations for vLLM==0.5.4

### CODE REVIEW

File: src/vllm_tgis_adapter/grpc/grpc_server.py [hunk 1]
```diff
@@ -199,14 +200,15 @@ def __init__(
...
-        self.config = await self.engine.get_model_config()
+        self.config = await self.client.get_model_config()
 
-        # Swap in the special TGIS stats logger
-        tgis_stats_logger = TGISStatLogger(
-            vllm_stat_logger=self.engine.engine.stat_loggers["prometheus"],
-            max_sequence_len=self.config.max_model_len,
-        )
-        self.engine.engine.stat_loggers["prometheus"] = tgis_stats_logger
+        # FIXME: figure out how to use the TGIS stats logger using the async engine client
```
Comment (dtrifiro) [inline src/vllm_tgis_adapter/grpc/grpc_server.py:205]: @maxdebayser This is what I think is going to be tricky to get working again

File: tests/conftest.py [hunk 2]
```diff
@@ -1,59 +1,58 @@
...
 
-import asyncio
 import sys
-import threading
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Annotated, TypeVar
 
 import pytest
-import requests
+import pytest_asyncio
```
Comment (dtrifiro) [inline tests/conftest.py:7]: Tests are currently broken. I was experimenting with `pytest_asyncio` to have async fixtures  that can be used to spin up an engine for the whole pytest session, but this is proving to be quite tricky.

File: src/vllm_tgis_adapter/__main__.py [hunk 3]
```diff
@@ -328,41 +61,30 @@ async def run_http_server(
...
+            task.cancel()
 
-    async def run() -> None:
+    async def override_signal_handler() -> None:
         loop = asyncio.get_running_loop()
-
-        http_server_task = loop.create_task(run_http_server(engine, args, model_config))
-        grpc_server_task = loop.create_task(
-            run_grpc_server(
-                engine, args, disable_log_stats=engine_args.disable_log_stats
```
Comment (dtrifiro) [inline src/vllm_tgis_adapter/__main__.py:354]: Note that I've gotten rid of the `disable_log_stats` argument, as the `_force_log` coroutine is scheduled as part of the `lifespan` attribute of the fastapi server.

File: pyproject.toml [hunk 4]
```diff
@@ -82,7 +83,7 @@ vllm_tgis_adapter = [
 ]
 
 [tool.pytest.ini_options]
-addopts = "-ra -W error::grpc.experimental.ExperimentalApiWarning"
+addopts = "-ra"
```
Comment (dtrifiro) [inline pyproject.toml:86]: `pytest -W` matches the behaviour of the python cli:

```
The full form of argument is: action:message:category:module:lineno
```

This caused `grpc.experimental.ExperimentalApiWarning` to be raised as an exception when running the healthcheck against the grpc server. We really don't need this.

File: src/vllm_tgis_adapter/__main__.py [hunk 5]
```diff
@@ -328,41 +61,30 @@ async def run_http_server(
...
-        loop.add_signal_handler(signal.SIGTERM, signal_handler)
-
-        await asyncio.gather(grpc_server_task, http_server_task)
-
-    event_loop.run_until_complete(run())
+        while not loop._signal_handlers:  # noqa: SLF001
+            logger.debug("No signal handlers are defined, sleeping")
+            await asyncio.sleep(1)
+
+        logger.debug("Signal handlers have been setup, proceeding with override")
```
Comment (dtrifiro) [inline src/vllm_tgis_adapter/__main__.py:78]: This trick is required because the signal handlers are set by `vllm.entrypoints.launcher.serve_http`, which are set up at a later stage: we have to wait for them to be set up in order to ovveride them and cleanly shut down the servers.

File: src/vllm_tgis_adapter/grpc/grpc_server.py [hunk 6]
```diff
@@ -876,25 +878,15 @@ async def ModelInfo(
...
-    # Log memory summary after model is loaded
-    from torch.cuda import memory_summary
-
-    assert isinstance(engine, AsyncLLMEngine)
-    assert isinstance(engine.engine, _AsyncLLMEngine)
-
-    if (device_type := engine.engine.device_config.device.type) == "cuda":
-        logger.info(memory_summary(engine.engine.device_config.device))
-    else:
-        logger.warning("Cannot print device usage for device type: %s", device_type)
```
Comment (dtrifiro) [inline src/vllm_tgis_adapter/grpc/grpc_server.py:890]: We cannot print the memory summary anymore, as this method is not available from the engine client.

File: src/vllm_tgis_adapter/grpc/grpc_server.py (hunk 1 above)
Comment (dtrifiro) [reply src/vllm_tgis_adapter/grpc/grpc_server.py]: We'll fix this in a later PR

File: src/vllm_tgis_adapter/grpc/grpc_server.py [hunk 7]
```diff
@@ -199,14 +199,16 @@ def __init__(
...
-        )
-        self.engine.engine.stat_loggers["prometheus"] = tgis_stats_logger
+        # FIXME: figure out how to use the TGIS stats logger using the
+        # async engine client
+        # # Swap in the special TGIS stats logger
+        # tgis_stats_logger = TGISStatLogger(
+        #     vllm_stat_logger=self.engine.engine.stat_loggers["prometheus"],
+        #     max_sequence_len=self.config.max_model_len,
+        # )
+        # self.engine.engine.stat_loggers["prometheus"] = tgis_stats_logger
```
Comment (njhill) [inline src/vllm_tgis_adapter/grpc/grpc_server.py:211]: Can we keep this and wrap it in:
```python
if isinstance(self.client, AsyncLLMEngine):
   ...
else:
    logger.warning("TGIS Metrics currently disabled in decoupled front-end mode, set DISABLE_FRONTEND_MULTIPROCESSING=True to enable")
```

File: src/vllm_tgis_adapter/grpc/grpc_server.py [hunk 8]
```diff
@@ -950,22 +942,26 @@ async def start_grpc_server(
...
-    disable_log_stats: bool,
-) -> None:
-    assert args is not None
+async def run_grpc_server(args: argparse.Namespace) -> None:
+    from vllm.entrypoints.openai.api_server import AsyncEngineRPCClient
 
-    server = await start_grpc_server(engine, args)
+    rpc_server_port = vllm.envs.VLLM_RPC_PORT
+    logger.info("gRPC server connecting to async engine on port %s", rpc_server_port)
+    async_engine_client = AsyncEngineRPCClient(rpc_server_port)
```
Comment (njhill) [inline src/vllm_tgis_adapter/grpc/grpc_server.py:950]: I think we need to share the `AsyncEngineClient` from the http server because it might not be an instance of `AsyncEngineRPCClient` .. if running in local mode

File: src/vllm_tgis_adapter/grpc/grpc_server.py [hunk 9]
```diff
@@ -950,22 +942,26 @@ async def start_grpc_server(
...
-            if not disable_log_stats:
-                await engine.do_log_stats()
     except asyncio.CancelledError:
         print("Gracefully stopping gRPC server")  # noqa: T201
+    except Exception:
+        logger.exception("Engine is dead")
+        # FIXME: handle this case?
+    finally:
         await server.stop(30)  # TODO configurable grace
         await server.wait_for_termination()
```
Comment (njhill) [inline src/vllm_tgis_adapter/grpc/grpc_server.py:967]: Again not critical but seems like there should be a cleaner way to do this

File: src/vllm_tgis_adapter/__main__.py (hunk 5 above)
Comment (njhill) [reply src/vllm_tgis_adapter/__main__.py:78]: I'm not sure about the way that async and non-async is mixed here, but it if works that's fine, we can always refactor it a bit later.

**Review by njhill — COMMENTED** [review]
Thanks @dtrifiro! It's nice that we can remove so much of the duplicated code now

File: src/vllm_tgis_adapter/grpc/grpc_server.py (hunk 9 above)
Comment (dtrifiro) [reply src/vllm_tgis_adapter/grpc/grpc_server.py:967]: The `except Exception` clause was a leftover from an older revision, I'll clean it up in the next push

File: tests/conftest.py [hunk 10]
```diff
@@ -81,13 +76,13 @@ def args(
...
-    monkeysession.setattr(
+    monkeypatch.setattr(
         sys,
         "argv",
         [
             "__main__.py",
-            f"--grpc-port={grpc_server_thread_port}",
-            f"--port={http_server_thread_port}",
+            f"--grpc-port={grpc_server_port}",
+            f"--port={http_server_port}",
```
Comment (maxdebayser) [inline tests/conftest.py:85]: Should we add `"--disable-frontend-multiprocessing"` for now? If there is an exception in the initialization of the server, the tests hang forever even with the timeout.

I noticed this because I sometimes set `NUM_GPUS=0`  and this causes a division by 0 division by 0 error in our tgis adapter code. (But this error is out of scope of this discussion)

Comment (maxdebayser) [reply tests/conftest.py:85]: FYI: https://github.com/vllm-project/vllm/issues/7213

File: src/vllm_tgis_adapter/__main__.py [hunk 11]
```diff
@@ -328,41 +61,30 @@ async def run_http_server(
...
-        event_loop = None
-
-    if event_loop is not None and event_loop.is_running():
-        # If the current is instanced by Ray Serve,
-        # there is already a running event loop
-        model_config = event_loop.run_until_complete(engine.get_model_config())
-    else:
-        # When using single vLLM without engine_use_ray
-        model_config = asyncio.run(engine.get_model_config())
+    loop = asyncio.new_event_loop()
```
[the remark this answers is not in the document]
Comment (dtrifiro) [reply src/vllm_tgis_adapter/__main__.py:88]: Done

File: src/vllm_tgis_adapter/grpc/grpc_server.py (hunk 7 above)
Comment (dtrifiro) [reply src/vllm_tgis_adapter/grpc/grpc_server.py]: Done

File: src/vllm_tgis_adapter/grpc/grpc_server.py (hunk 8 above)
Comment (dtrifiro) [reply src/vllm_tgis_adapter/grpc/grpc_server.py]: Done

File: src/vllm_tgis_adapter/grpc/grpc_server.py (hunk 9 above)
Comment (dtrifiro) [reply src/vllm_tgis_adapter/grpc/grpc_server.py]: Cleaned it up

File: tests/conftest.py (hunk 10 above)
Comment (dtrifiro) [reply tests/conftest.py:99]: I cleaned up the tests code a bit, so you might not see that hang anymore.

Regarding `--disable-frontend-multiprocessing`: I parametrized tests so that we test both enabled and disabled.

Tests pass in both cases, although there's an annoying bug at tests cleanup which causes pytest to exit with a non-zero exit code even with all tests passing. I ended up temporarily disabling that test so that we can get this merged asap.

Comment (maxdebayser) [reply tests/conftest.py:99]: FYI, https://github.com/vllm-project/vllm/pull/7217 fixes this particular hang.

File: src/vllm_tgis_adapter/grpc/grpc_server.py (hunk 1 above)
Comment (dtrifiro) [reply src/vllm_tgis_adapter/grpc/grpc_server.py]: Actually fixed it in this one. See newer versions

**Comment by dtrifiro** [comment]
Parametrization is disabled because of [this](https://github.com/opendatahub-io/vllm-tgis-adapter/pull/84#issuecomment-2273175910) error
